GOALS.md

```diff
@@ -0,0 +1,233 @@
+like a scheduler or executor of sorts for the various plugins. Because it knows the state of all the plugins running,
+as well as their (potentially complex) I/O configuration, it can apply massive optimizations to its scheduling and
+processing, such as spreading the work on a single thread pool. However, this means the plugins must cooperate, and
+cannot do what they want. Examples of work LV2 plugins should defer to the host are I/O buffers and communication, state
+serialization (i.e., presets / session loading and saving), asynchronous processing, or UI communication.
+
+In that aspect, LV2 plugins and hosts share very similar behaviors with Rust's own
+[Futures](https://doc.rust-lang.org/std/future/trait.Future.html) and
+[Executors](https://docs.rs/futures/0.3.16/futures/executor/index.html).
+
+Like with Rust Futures, the execution system for LV2 plugins is inherently *cooperative*. LV2 plugins *must* finish
+processing before another plugin can run. If a plugin, just like a Future, uses blocking operations, it will block
+the whole thread without any means of interruption by the host. This includes (but is not limited to): memory
+allocations, multi-thread processing, thread synchronization (atomics, locks), general I/O, and more.
+
+If a plugin needs to perform asynchronous work for instance (like loading and decoding a sample file), they should use
+the [LV2 Worker](http://lv2plug.in/ns/ext/worker) API (see the `lv2-worker` crate). Just like a Future would use the
+executor's [`spawn()`](https://docs.rs/futures/0.3.16/futures/task/trait.SpawnExt.html#method.spawn) method to process
+additional asynchronous work, instead of doing it synchronously or spawning a thread of its own.
+
+## Extra ergonomics and sugar
+Also, while ergonomics and extra utilities are nice to have in this library sometimes, they *must* be optional to use.
+Indeed, because of the goal to be a low-level library, we must not prevent the user from doing custom things themselves
+at the cost of complexity. At least, as long as it is safe for them to do so.
+
+A good example of this is the `UridCollection` derive macro in the `urid` crate. While users can make similar
+collections manually using 100% safe and sound code, it is a very tedious and boilerplate-heavy implementation that
+can be abstracted away. This is thanks to the fact that `URID<T>` is integrated to the type system, allowing the
+structure to be manually filled if desired.
+
+An anti-example of this is the `PortCollection` derive macro in the `core` crate. It may seem similar to
+`UridCollection` at a first glance (it is a collection of things that can be filled automatically), but there is a
+catch. An invalid `index <-> port` mapping implementation means that bad pointer casts are going to be made, and this
+is definitely Undefined Behavior.
+
+In this case, this ergonomic helper is necessary to produce safe code, and is not considered optional: the
+`PortCollection` abstraction cannot be bypassed.
```

Comment:
I think the `PortCollection` abstraction can in fact be bypassed: it can be implemented outside of the `lv2` crate. For [`rsynth`](https://crates.io/crates/rsynth), this is in fact desirable, otherwise it's hard to abstract over different API's.

Reply:
I think this is an unclear wording issue. The `PortCollection` trait itself cannot be bypassed: the `Plugin` trait (among others) rely heavily on it. However, the `PortCollection` derive macro *can* be bypassed, and you can implement `PortCollection` however you like outside of the `lv2` crate, which, as you say, is useful if you want to make your own abstraction.

In that specific case, I consider the `PortCollection` trait to be necessary for safe operation, and the `PortCollection` derive macro to be the optional syntax sugar layer on top (like all derive macros in `rust-lv2` are).

Will fix this as well. :slightly_smiling_face: